app/models/package_manager/base/version_updater.rb

```diff
@@ -49,7 +51,8 @@ def upsert_version_for_project!
       private
 
       def db_project_version
-        @db_project_version ||= @project.versions.find_or_initialize_by(number: @api_version_to_upsert.version_number)
+        @db_project_version ||= @preloaded_db_versions.find { |v| v.number == @api_version_to_upsert.version_number } ||
+                                @project.versions.new(number: @api_version_to_upsert.version_number)
```

Comment:
here's the optimization: instead of doing a `SELECT * FROM versions LIMIT 1` for each API version, we're now searching a preloaded list of versions in ruby-land.